modules/PatternUtils.js

```diff
@@ -167,23 +167,60 @@ export function formatPattern(pattern, params) {
       if (paramValue != null)
         pathname += encodeURI(paramValue)
     } else if (token === '(') {
+      parenHistory[parenCount] = ''
       parenCount += 1
     } else if (token === ')') {
+      const parenText = parenHistory.pop()
       parenCount -= 1
+
+      if (parenCount)
+        parenHistory[parenCount - 1] += parenText
+      else
+        pathname += parenText
     } else if (token.charAt(0) === ':') {
       paramName = token.substring(1)
       paramValue = params[paramName]
 
       invariant(
         paramValue != null || parenCount > 0,
-        'Missing "%s" parameter for path "%s"',
+        'missing "%s" parameter for path "%s"',
         paramName, pattern
       )
 
-      if (paramValue != null)
+      if (paramValue == null) {
+        if (parenCount) {
+          parenHistory[parenCount - 1] = ''
+
+          const curTokenIdx = tokens.indexOf(token)
+          const tokensSubset = tokens.slice(curTokenIdx, tokens.length)
+          let nextParenIdx = -1
+
+          for (let i = 0; i < tokensSubset.length; i++) {
+            if (tokensSubset[i] == ')') {
+              nextParenIdx = i
+              break
+            }
+          }
+
+          invariant(
+            nextParenIdx > 0,
+            'missing end parameter'
```

Comment:
This is still unfriendly to the end-user. I would include some context to point them in the right direction. E.g., `'Path "%s" is missing end parameter at segment "%s"', pattern, tokensSubset.join('')`